### PR TITLE
CytoPipe CLI implementation

### PR DESCRIPTION
**Pull request by axiomcura**
## Motivation

`Cytopipe` is a going to be a python package that will contains extensive amount of workflows to processing cell morphology features from both [CellProfiler](https://github.com/CellProfiler/CellProfiler) and [DeepProfiler](https://github.com/cytomining/DeepProfiler). 

This PR will introduces the very first working CLI interface for `Cytopipe`!

## Diagram 

Below is a simple diagram on how the CLI implementation is designed.

Cytopipe-cli_PR drawio

## Usage

The `CLI` interface is easily executed by typing:

```
cytopipe <MODE> <MODE AGURMENTS>
```

The `MODE` is an argument that tells `cytopipe`'s what type of action the user is performing. Each mode will serve a different purpose.

Currently, `cytopipe` contains three modes 

- `init` → initializes required files into the appropriate format (files, file directories, etc)
- `run` → executing workflows
- `help` → displays CLI usage documentation

### Example

The first thing to do before trying to execute `cytopipe`'s workflows is to format the files that are required for the processing. 

This is done by using the `init` mode:

```
# example `[ ] ` indicates the parameter takes a list
# `< > ` indicates a string type
cytopipe init -d [ data1.sqlite, data2.sqlite ..], -m <metadata_dir> -p <platemap_path> -b <barcodes_path>
```

Once the files are appropriately formatted, the next step is to use the `run` mode and pick the desired workflow to process your data 

```
cytopipe run <workflow_name>
```

**work in progress:** 

To see what workflows are available, you can use the `help` option under the `run` mode and it will list all available workflows:

```
cytopipe run help
```

**Note:** This PR focuses on implementing the a work CLI for `cytopipe` ; however, more features will be developed in the future! 

A future feature includes a `test` mode, which tests all the functionality of `cytopipe` before conducting any processes. 

This includes checking the source code, software stack and versions, which is essential for both developers and users for debugging purposes!

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from setuptools import setup, find_packages
```
Comment (gwaybio) [inline setup.py:1]: This probably belongs in a separate PR, as it is the initiating steps to pip install. OK to have here though (just link to this PR when making the official PIP implementation, and maybe change name of this PR to include setup.py)

**Review by gwaybio — APPROVED** [review]
LGTM, only really minor, surface level comments. One general comment is about docstring formatting (i'm used to seeing `"""`)

**Comment by gwaybio** [comment]
oh! and we should somehow try to capture the documentation you wrote for this PR description somewhere more visible (perhaps in the main readme!)

File: cytopipe/cli/cli_docs.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+Required Arguments:
+    workflow            Name of the workflow to execute
+
+Optional Arguments:
+    -c, --max_cores     Maximum number of cores to use for the workflow 
+                        default=1
+
+Help Arguments:
+    help                Displays cytopipe's Run mode documentation j
```
[the remark this answers is not in the document]
Comment (axiomcura) [reply cytopipe/cli/cli_docs.py:21]: Yes this is a typo!

File: cytopipe/cli/args.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    # CLI inputs
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "mode",
+        choices=["init"],
+        help="Init mode sets up the all files for cytopipe processing",
+    )
+    parser.add_argument(
+        "-d", "--data", nargs="+", required=True, help="list of plate data files"
```
[the remark this answers is not in the document]
Comment (axiomcura) [reply cytopipe/cli/args.py:48]: >this has the potential to get ugly (e.g. what happens if someone has 1,000 plates to process?). 
I think it is ok as-is now, for prototyping. But we should consider being able to pass a file into the --data flag.

In this senario, 1,000 plates means 1,000 plates along with it's metadata  and bar code files, correct? That should not be a problem as the `--data` parameter takes in a list of plates. This is denoted by the `nargs="+"`. 

For example, users can use wildcards to capture all 1,000s plates:
```
cytopipe init -d *_plate.sqlite -m metdata_dir/ -b barcodes.txt
```

Also, 
>ohh, I see. metadata and barcode are separate flags. ￼ maybe data should be an optional parameter then? To be used only if you want to process a portion of the data?

I ended up removing the `-p` option because I realize that platemaps are stored within the metadata directory. 

This is the commit: https://github.com/WayScience/CytoPipe/pull/15/commits/d4ef1fdf05febe35c7f8a5c9703a210bb5066d50


Hopefully this answers your question.

File: generate_sample_inputs.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import os
```
[the remark this answers is not in the document]
Comment (axiomcura) [reply generate_sample_inputs.py:1]: Ah! forgot to put that in the .gitignore! But putting it in the test suite is not a bad idea :thinking:

File: setup.py (hunk 1 above)
Comment (axiomcura) [reply setup.py:1]: Ahhh, This was not my attempt to make `cytopipe` pip installable. In fact, the entry points (how a user communicates with `cytopipe`) must be explicitly written in a `setup.py` file. [source](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#console-scripts)

But yes, in the future, making `cytopipe` pip installable will be a separate PR!

Here's the code block: https://github.com/WayScience/CytoPipe/pull/15/files/d4ef1fdf05febe35c7f8a5c9703a210bb5066d50#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7

File: cytopipe/cli/args.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+# ------------------------------------------------------------
+# args.py
+#
+# Responsible for centralizing and handling all user inputted
+# parameters for cytopipe.
+#
+# Contains different modes with their respected parameters
+# ------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (axiomcura) [reply cytopipe/cli/args.py:8]: I didn't know there was a standard why of doing it. I will switch it to `"""`

**Comment by axiomcura** [comment]
Currently working on that! Will be a separate PR soon!

File: cytopipe/cli/args.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwaybio) [reply cytopipe/cli/args.py]: https://peps.python.org/pep-0257/
